forgerock-openbanking-uk-aspsp-rs/forgerock-openbanking-uk-aspsp-rs-gateway/forgerock-openbanking-uk-aspsp-rs-gateway-server/src/main/java/com/forgerock/openbanking/aspsp/rs/api/payment/verifier/DetachedJwsVerifier.java

```diff
@@ -104,22 +112,26 @@ private boolean isB64ClaimHeaderSetToFalse(String jwsDetachedSignature, String p
         String jwsSerialized = rebuildJWS(jwsDetachedSignature, payload);
         log.debug("The reconstructed JWS from the detached signature: {}", jwsSerialized);
         SignedJWT jws = (SignedJWT) JWTParser.parse(jwsSerialized);
-        if (jws.getHeader().getCustomParam("b64") != null) {
-            log.debug("B64 claim header in the detached signature: {}", jws.getHeader().getCustomParam("b64"));
-            Object b64Param = jws.getHeader().getCustomParam("b64");
-            if (!(b64Param instanceof Boolean)) {
-                return false;
-            }
-            return ((Boolean)b64Param) == false;
-        }
-        return false;
+        boolean isSetToFalse = jws.getHeader().isBase64URLEncodePayload();
+        return !isSetToFalse;
     }
 
     private boolean isB64ClaimHeaderPresent(String jwsDetachedSignature, String payload) throws OBErrorException, ParseException {
         String jwsSerialized = rebuildJWS(jwsDetachedSignature, payload);
         log.debug("The reconstructed JWS from the detached signature: {}", jwsSerialized);
         SignedJWT jws = (SignedJWT) JWTParser.parse(jwsSerialized);
-        return jws.getHeader().getCustomParam("b64") != null;
+        Set<String> params = jws.getHeader().getIncludedParams();
+        return params.contains("b64");
+    }
+
+
+    private boolean newIsB64ClaimHeaderPresent(String jwsDetachedSignature, String payload) throws OBErrorException {
```

Comment:
Was this intended to replace the other method?

Reply:
Ah yes. Do you approve? of the approach?

Reply:
Based on what you were saying, it will cover the scenario where b64 is included, but not declared as a critical parameter. However, I don't think that's valid according to [RFC-7797](https://tools.ietf.org/html/rfc7797):

> The "crit" Header Parameter MUST be included with "b64" in its set of values when using the "b64" Header Parameter to cause implementations not implementing "b64" to reject the JWS (instead of it being misinterpreted).

The only thing I'm not sure about is that your examining a String, rather than a valid JWS, which the Nimbus library will do.

Reply:
Otherwise, it's fine if you think my implementation is too naive:

```
private boolean isB64ClaimHeaderPresent(String jwsDetachedSignature, String payload) throws OBErrorException, ParseException {
   String jwsSerialized = rebuildJWS(jwsDetachedSignature, payload);
   log.debug("The reconstructed JWS from the detached signature: {}", jwsSerialized);
   SignedJWT jws = (SignedJWT) JWTParser.parse(jwsSerialized);
   return jws.getHeader().getCriticalParams().contains("b64");
}
```

Reply:
If I thought Lloyds would actually read RFC-7797 and build their tests according to that spec then I'd agree your code is fine. However, I don't trust them to do that. I fully expect them to just add b64: true to a JWS and try it....